hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/exec/systemcontracts/hts/updatetokencustomfees/UpdateTokenCustomFeesDecoder.java

```diff
@@ -0,0 +1,186 @@
+                .build());
+    }
+
+    private Stream<CustomFee> decodeFractionalFees(
+            @NonNull final Tuple[] fractionalFees, @NonNull final AddressIdConverter addressIdConverter) {
+        return Arrays.stream(fractionalFees).map(fee -> CustomFee.newBuilder()
+                .fractionalFee(FractionalFee.newBuilder()
+                        .fractionalAmount(Fraction.newBuilder()
+                                .numerator(fee.get(FRACTIONAL_FEE_NUMERATOR))
+                                .denominator(fee.get(FRACTIONAL_FEE_DENOMINATOR))
+                                .build())
+                        .minimumAmount(fee.get(FRACTIONAL_FEE_MIN_AMOUNT))
+                        .maximumAmount(fee.get(FRACTIONAL_FEE_MAX_AMOUNT))
+                        .netOfTransfers(fee.get(FRACTIONAL_FEE_NET_OF_TRANSFERS))
+                        .build())
+                .feeCollectorAccountId(addressIdConverter.convert(fee.get(FRACTIONAL_FEE_FEE_COLLECTOR)))
+                .build());
+    }
+
+    private Stream<CustomFee> decodeRoyaltyFees(
+            @NonNull final Tuple[] royaltyFees, @NonNull final AddressIdConverter addressIdConverter) {
+        return Arrays.stream(royaltyFees).map(fee -> CustomFee.newBuilder()
+                .royaltyFee(RoyaltyFee.newBuilder()
+                        .exchangeValueFraction(Fraction.newBuilder()
+                                .numerator(fee.get(ROYALTY_FEE_NUMERATOR))
+                                .denominator(fee.get(ROYALTY_FEE_DENOMINATOR))
+                                .build())
+                        .fallbackFee(getFallbackFee(fee))
+                        .build())
+                .feeCollectorAccountId(addressIdConverter.convert(fee.get(ROYALTY_FEE_FEE_COLLECTOR)))
+                .build());
+    }
+
+    // In the solidity structure for Fixed Fees we have the property `bool useCurrentTokenForPayment` that is not
+    // present in the protobuf version.
+    // In order to determine the denominating token we need to check if the property is present.
+    private TokenID determineDenominatingToken(@NonNull final Tuple fee, @NonNull final TokenID tokenId) {
+        final boolean useCurrentToken = fee.get(FIXED_FEE_USE_CURRENT_TOKEN_FOR_PAYMENT);
+        return useCurrentToken
+                ? tokenId
+                : getIfPresent(fee.get(FIXED_FEE_TOKEN_ID), fee.get(FIXED_FEE_USE_HBARS_FOR_PAYMENT));
+    }
+
+    private TokenID getIfPresent(@NonNull final Address address, final boolean useHbarsForPayment) {
+        final var tokenId = ConversionUtils.asTokenId(address);
+        return useHbarsForPayment || tokenId.equals(TokenID.DEFAULT) ? null : tokenId;
+    }
+
+    private FixedFee getFallbackFee(@NonNull Tuple fee) {
+        final Address tokenAddress = fee.get(ROYALTY_FEE_TOKEN_ID);
+        final long amount = fee.get(ROYALTY_FEE_AMOUNT);
+        return ConversionUtils.asTokenId(tokenAddress).equals(TokenID.DEFAULT) && amount == 0
```

Comment:
The expression `ConversionUtils.asTokenId(tokenAddress).equals(TokenID.DEFAULT)` really comes down to having  `tokenAddress.value().equals(BigInteger.ZERO)` right?  Just checking my understanding.

This expression here does a teeny bit more work but I think it expresses the semantics better.